lib/src/kbasesearchengine/events/handler/WorkspaceEventHandler.java

```diff
@@ -649,4 +681,140 @@ private static String toWSRefPath(final List<GUID> objectRefPath) {
         return String.join(";", refpath);
     }
 
+    private final StatusEvent updateEventForDeletion(final StatusEvent ev)
+              throws RetriableIndexingException {
+        // wsid and objid of the object that the specified StatusEvent is an event for
+        final int wsid = ev.getAccessGroupId().get();
+        final long objid = Long.valueOf(ev.getAccessGroupObjectId().get()).longValue();
+
+        // check if workspace is permanently deleted or marked as deleted
+        try {
+            getWorkspaceInfo(wsid);
+        } catch (IOException ex) {
+            handleException(ex);
+        } catch (JsonClientException ex) {
+            // hacky: if exception message is changed, this logic will silently fail
+            if (ex.getMessage().contains("No workspace with id ")) {
+                return StatusEvent.
+                        getBuilder(ev.getStorageObjectType().get(), ev.getTimestamp(),
+                                StatusEventType.DELETE_ALL_VERSIONS).
+                        withNullableAccessGroupID(wsid).
+                        withNullableObjectID(Long.toString(objid)).build();
+            }
+            else {
+                handleException(ex);
+            }
+        }
+
+        // check if object is permanently deleted or marked as deleted
+        final Map<String, Object> command;
+        command = new HashMap<>();
+        command.put("command", "listObjects");
+        command.put("params", new ListObjectsParams().
+                withIds(Arrays.asList((long)wsid)).
+                withMinObjectID(objid).
+                withMaxObjectID(objid));
+        try {
+            boolean objidExists = false;
+
+            final List<List> objList = ws.getClient().administer(new UObject(command))
+                    .asClassInstance(List.class);
+
+            if(objList != null && objList.size() > 0) {
+                objidExists = true;
+            }
+
+            if (!objidExists) {
```

Comment:
Could simplify here by getting rid of objidExists and just doing

```
if (!objList.isEmpty()) {
```
Since objList is guaranteed to never be null

Reply:
Or if objidExists makes things more readable IYO:

```
final boolean objidExists = !ws.getClient().administer(new UObject(command))
                    .asClassInstance(List.class).isEmpty();
```

Reply:
Made the change.